File: 2019/Open/3_fenceplan/fenceplan.cpp
#include <iostream>
#include <vector>
#include <queue>
#include <algorithm>
using namespace std;

typedef long long ll;

vector<ll> x;
vector<ll> y;
int N, M;

int dfs(vector<vector<int>> adj, vector<bool> visited) {
  ll minPerimeter = 9223372036854775807;

  for (int i = 1; i < N + 1; i++) {
    if (!visited[i]) {
      // run dfs
      queue<int> q;
      q.push(i);

      ll minX = 100000001;
      ll minY = 100000001;
      ll maxX = -100000001;
      ll maxY = -100000001;

      while (!q.empty()) {
        int cur = q.front();
        q.pop();
        visited[cur] = true;

        minX = min(minX, x[cur]);
        maxX = max(maxX, x[cur]);
        minY = min(minY, y[cur]);
        maxY = max(maxY, y[cur]);

        for (int a : adj[cur]) {
          if (!visited[a]) {
            q.push(a);
          }
        }
      }

      minPerimeter = min(minPerimeter, (maxX - minX) * 2 + (maxY - minY) * 2);
    }
  }

  return minPerimeter;
}

int main() {
  freopen("fenceplan.in", "r", stdin);
  freopen("fenceplan.out", "w", stdout);

  vector<vector<int>> adj;
  vector<bool> visited;

  cin >> N >> M;
  for (int i = 0; i < N + 1; i++) {
    x.push_back(0);
    y.push_back(0);
    visited.push_back(false);
    vector<int> tmp;
    tmp.shrink_to_fit();
    adj.emplace_back(move(tmp));
  }

  ll a, b;
  for (int i = 1; i < N + 1; i++) {
    cin >> a >> b;
    x[i] = a;
    y[i] = b;
  }

  int c, d;
  for (int i = 0; i < M; i++) {
    cin >> c >> d;

    adj[c].push_back(d);
    adj[d].push_back(c);
  }

  cout << dfs(adj, visited) << endl;

  return 0;
}

File: 2019/Open/3_fenceplan/Fenceplan.java
import java.io.PrintWriter;
import java.util.*;

public class Fenceplan {
  static int N;
  static int M;

  public static void main(String[] args) throws Exception {
    Scanner sc = new Scanner(new java.io.File("fenceplan.in"));
    PrintWriter pw = new PrintWriter("fenceplan.out");
    // Scanner sc = new Scanner(System.in);

    N = sc.nextInt();
    M = sc.nextInt();

    Cow[] cows = new Cow[N];
    for (int i = 0; i < N; i++) {
      cows[i] = new Cow(sc.nextInt(), sc.nextInt());
    }

    for (int i = 0; i < M; i++) {
      int a = sc.nextInt()-1;
      int b = sc.nextInt()-1;
      cows[a].neighbors.add(cows[b]);
      cows[b].neighbors.add(cows[a]);
    }

    int minPerimeter = Integer.MAX_VALUE;

    for (int i = 0; i < N; i++) {
      if (cows[i].visited) continue;

      // bfs
      Queue<Cow> q = new LinkedList<>();
      q.add(cows[i]);
      cows[i].visited = true;

      int maxX = 0;
      int maxY = 0;
      int minX = Integer.MAX_VALUE;
      int minY = Integer.MAX_VALUE;


      while (!q.isEmpty()) {
        Cow curCow = q.poll();
        maxX = Math.max(maxX, curCow.x);
        maxY = Math.max(maxY, curCow.y);
        minX = Math.min(minX, curCow.x);
        minY = Math.min(minY, curCow.y);

        for (Cow c : curCow.neighbors) {
          if (!c.visited) {
            q.add(c);
            c.visited = true;
          }
        }

      }
      minPerimeter = Math.min(minPerimeter, 2*(maxX-minX)+2*(maxY-minY));
    }

    pw.println(minPerimeter);

    pw.close();
    sc.close();
  }

  public static class Cow {
    int x;
    int y;
    boolean visited = false;
    List<Cow> neighbors = new ArrayList<>();

    public Cow (int x, int y) {
      this.x = x;
      this.y = y;
    }
  }
}
